### PR TITLE
measuring startup

### PR DESCRIPTION
**Pull request by cfbolz**
the goal is to retain all the krun features but without having to change
krun itself in major ways. startup.krun documents the patching that it
does to achieve that.

### CODE REVIEW

File: startup.krun [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    "default-ruby": EntryPoint("bench.rb", subdir="ruby"),
+    "default-javascript": EntryPoint("bench.js", subdir="javascript"),
+}
+
+ITERATIONS_ALL_VMS = 1
+
+def patch_runner(vm_def):
+    from krun.env import EnvChangeAppend
+    if vm_def.iterations_runner == "IterationsRunner":
+        #import pdb; pdb.set_trace()
```
Comment (vext01) [inline startup.krun]: debugging code

File: startup.krun [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    return vm_def
+
+from krun.platform import DebianLinuxPlatform, BasePlatform
+
+def bench_cmdline_adjust(self, args, env_dct):
+    prepend_args = BasePlatform.bench_cmdline_adjust(self, [], env_dct)
+    if "check_user_change" in args[-3]:
+        return prepend_args + args
+    return prepend_args + ["krun/iterations_runners/startup_runner_runner_c"] + args
+DebianLinuxPlatform.bench_cmdline_adjust = bench_cmdline_adjust
```
Comment (vext01) [inline startup.krun]: Does this only work on Debian, or is the approach (in principal) platform specific? I imagine Laurie will want to measure startup on OpenBSD.

File: startup.krun [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        assert "iterations_runner." in vm_def.iterations_runner
+        _, suffix = vm_def.iterations_runner.rsplit(".", 1)
+        vm_def.iterations_runner = os.path.join(STARTUP_RUNNER_DIR, "startup_runner." + suffix)
+    return vm_def
+
+from krun.platform import DebianLinuxPlatform, BasePlatform
+
+def bench_cmdline_adjust(self, args, env_dct):
+    prepend_args = BasePlatform.bench_cmdline_adjust(self, [], env_dct)
+    if "check_user_change" in args[-3]:
```
Comment (vext01) [inline startup.krun]: Why do we need this? Can we add a comment here?

File: startup.krun [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        return prepend_args + args
+    return prepend_args + ["krun/iterations_runners/startup_runner_runner_c"] + args
+DebianLinuxPlatform.bench_cmdline_adjust = bench_cmdline_adjust
+
+VMS = {
+        'C': {
+                'vm_def': patch_runner(NativeCodeVMDef()),
+                'variants': ['default-c'],
+                'n_iterations': ITERATIONS_ALL_VMS,
+                'warm_upon_iter': 0,
```
Comment (vext01) [inline startup.krun:99]: AFAIK, `warm_upon_iter` is not used. I have been meaning to remove all traces of this from the krun repo.

File: startup.krun [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    #"*:Hotspot:*",
+    #"*:Graal:*",
+    #"*:LuaJIT:*",
+    #"*:HHVM:*",
+    #"*:JRubyTruffle:*",
+    #"*:V8:*",
+]
+
+N_EXECUTIONS = 200  # Number of fresh processes.
+N_GRAPHS_PER_BENCH = 2
```
Comment (vext01) [inline startup.krun:173]: This is also an unused config variable.

File: startup.krun [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+# this file is a minimally hacky way to do startup measurements with all the
+# krun goodies, but without changing krun.
+#
+#
+# the way it works is as follows:
+#
+# before the invocation of the VM, another executable is inserted:
+#
+# sudo -u krun nice -20 taskset 0x2 env <some variables> \
+# startup_runner_runner_c jruby startup_runner.rb
```
Comment (vext01) [inline startup.krun]: It took me a while to realise that there is a `startup_runner_c` and a `startup_runner_runner_c`. I wonder if we can find a better naming convention.

File: startup_runners/startup_runner.rb [hunk 7]
```diff
@@ -0,0 +1,15 @@
+if /linux/ =~ RUBY_PLATFORM
```
Comment (vext01) [inline startup_runners/startup_runner.rb]: This needs the same hack as we just applied to the regular `iterations_runner.rb`. I.e. on JRuby we should call to `clock_gettime_monotonic()` in `libkruntime`.

File: startup_runners/startup_runner_runner.c [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+/* To correctly expose asprintf() on Linux */
+#define _GNU_SOURCE
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <limits.h>
+#include <errno.h>
+#include <string.h>
+#include <dlfcn.h>
```
Comment (vext01) [inline startup_runners/startup_runner_runner.c]: Some of these includes are redundant. Certainly `dlfcn.h`, maybe `limits.h` and `string.h`.

File: startup.krun (hunk 6 above)
Comment (cfbolz) [inline startup.krun]: `outer_startup_runner_c`?

File: startup.krun (hunk 2 above)
Comment (cfbolz) [inline startup.krun]: yeah, should work, though the C file might need adaptations.

File: startup.krun (hunk 3 above)
Comment (cfbolz) [inline startup.krun]: will add one.

Comment (cfbolz) [inline startup.krun]: The basic ideas is that all the platform checks are run as benchmarks krun. That means their output needs to be valid JSON. This does not work with the startup_runner_runner_c for the check_user_change platform check.

File: startup.krun (hunk 4 above)
Comment (cfbolz) [inline startup.krun:99]: ok, it's copy-pasted from warmup.krun, will kill.

File: startup.krun (hunk 5 above)
Comment (cfbolz) [inline startup.krun:173]: ditto

File: startup_runners/startup_runner.rb (hunk 7 above)
Comment (cfbolz) [inline startup_runners/startup_runner.rb]: Good catch (missed a push from bencher3).

File: startup.krun (hunk 2 above)
Comment (vext01) [inline startup.krun]: OK, so I wonder if we can patch the right platform class automatically.  One idea: we use `detect_platform()` in `platform.py` then ask the return value for it's type. This is then the class we patch. What do you think?

File: startup.krun (hunk 4 above)
Comment (vext01) [inline startup.krun:99]: Yes, I've been slacking on that front. I will make a PR to eliminate them now.

File: startup.krun (hunk 2 above)
Comment (cfbolz) [inline startup.krun]: works for me, will do that

**Comment by vext01** [comment]
This all looks good to me. Do you want to do any squashing?

**Comment by cfbolz** [comment]
squashed

**Pull request merged**